keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+  [packages](#pkg-list) and [platforms](#plat-list)
+
+- [ ] The output of the above tool (a directory tree, holding a set of
+  generated packages) ideally is in a format that can be uploaded to a GCS
+  bucket which can be used as a debian package repo / yum repo.
+
+  If that is not feasible there needs to be a tool that can re-arrange that
+  directory and prepare it for upload. This might also give the opportunity to
+  run certain tools to prepare the directory, e.g. [`dpkg-scanpacakges`,
+  `debpool` or other).
+
+- [ ] Create one tool that does "everything" that needs to be done in the
+  release and can be called by [`anago`][anago].
+
+  Without any special configuration is should do whatever is needed so that it
+  can be called from within [`anago`][anago], however there should be options
+  to specify which components to package, which components to package, get the
+  artefacts by compiling them or pulling them from a staging bucket, ...
+  
+  This might be seen as the
+  "orchestrator tool" and has the following responsibilities:
+  1. determinate the version to be packaged, based on the current git checkout
+     ok `k/k` (a version like in `bazel-genfiles/build/version`)
+  1. do a build (`build/run.sh make $COMPONENT` or such) or download all needed
+     artefacts from from a staging bucket
+  1. run the tool to prepare packages for upload
+  1. upload the packages
+
+
+[anago]: https://github.com/kubernetes/release/blob/master/anago
+
+<!--
+### User Stories [optional]
+
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of the system.
+The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they releate.
+-->
+
+### Risks and Mitigations
+
+- *Risk*: creating another way on how to build packages  
+  *Mitigations*:  
+  - work with the current builder (@sumi) in google
+  - writing super extra great docs so this is the goto way on how to build
+- *Risk*: `nfpm` is not the right tool, it goes away, works differently, ...  
+  *Mitigations*:  
+  - we encapsulate the build of exactly one package, if things change, we can
+    change the internal implementation but keep the external
+    interface
+- *Risk*: We cannot build proper windows packages but community wants/needs them  
+  *Mitigations*:
+  - Talk to people in sig windows, esp. about packaging their components
+
+## Design Details
+
+### Test Plan
+
+- different components will have unit tests as much as possible
+- comparing output debs/rpms to what currently is released
+  - not sure how? by the files they deploy?
```

Comment:
we could SHA the contents, yes.

also ideally we should have test jobs that build the artifacts and install them within containers on various distros, but this might take a while to become reality. windows and osx might be trickier as part of the e2e here.

as an initial step we might want to only cover debs and rpms per one distro each?

Reply:
test automation should use the sha'd packages for every build in CI.

Reply:
>> - comparing output debs/rpms to what currently is released
>>   - not sure how? by the files they deploy?
>
> we could SHA the contents, yes.

I was more thinking along the lines: Can we set up some test for quick feedback which compare packages that we currently release (via k/release) with the ones that were created with "the new way" without installing them, spinning up a cluster and running the k8s e2e tests.

As the packages will be different, sha'ing things is useless for that case.

I was thinking along the lines of a test that verifies that the "new" packages have at least the files the "old" packages have, determined by the file path. And maybe verifying that the "new" packages have at dependencies on at least the same packages as the "old" packages. Just for the transition phase to have some confidence that the "new" packages are not "worse" then the old packages. 

I am not convinced how useful those those tests would be so I will drop that.